Fix dock visibility issues

kitbdev opened the pull request:
Fix dock visibility for Editor Features and distraction free mode.

- fixes https://github.com/godotengine/godot/issues/66834
- fixes #84125 crash when changing Editor Features with floating docks (import, filesystem, node, or history)
- fixes error when closing floating dock if there are less tabs than before
- fixes loading layouts in distraction free mode being very buggy (bottom dock disappears, side docks might open and be empty)
- fixes floating dock still opened even if it was disabled in Editor Features when loading layout

Updated `TabContainer`'s `move_tab_from_tab_container` to preserve all tab data.

The crash fix is a workaround, ideally it wouldn't have to close all floating docks. I am working on a different pr that should fix this in the future.

There were 4 separate implementations to update the dock slots visibility:
`update_dock_slots_visibility()`,  `_update_dock_containers()`, in `_load_docks_from_config()`, and `_dock_tab_changed()` (removed the lines affecting bottom docks for this one, see #39177).
I have consolidated them all in `update_dock_slots_visibility()`.

kitbdev committed:
fix dock visibility issues

YeldhamDev commented on scene/gui/tab_container.cpp:
```diff
@@ -381,17 +381,23 @@ void TabContainer::move_tab_from_tab_container(TabContainer *p_from, int p_from_
 	p_from->remove_child(moving_tabc);
 	add_child(moving_tabc, true);
 
 	set_tab_title(get_tab_count() - 1, tab_title);
 	set_tab_icon(get_tab_count() - 1, tab_icon);
+	set_tab_button_icon(get_tab_count() - 1, tab_button_icon);
 	set_tab_disabled(get_tab_count() - 1, tab_disabled);
+	set_tab_hidden(get_tab_count() - 1, tab_hidden);
 	set_tab_metadata(get_tab_count() - 1, tab_metadata);
+	get_tab_bar()->set_tab_icon_max_width(get_tab_count() - 1, tab_icon_max_width);
```
Any reason why those are set before the child is moved to the right index?

kitbdev replied:
That's how it was done before, so I just kept it the same. I can change it to be after the move.

YeldhamDev replied:
Yes, as it would be preferable to use `p_to_index` instead of `get_tab_count() - 1` over and over.

kitbdev commented:
Since this fixes a crash, should it be for 4.2?

AThousandShips commented:
Can you please open an issue for the crash part, I don't see it reported in the linked issue, a crash is serious enough to track actively

AThousandShips commented:
But note also that we're in the late stage of the 4.2 cycle and focus right now is on bugs introduced in 4.2 and fixing them so 4.2 introduces as few *new* bugs as possible, and as some of this behaviour goes back to 4.0 it's lower priority right now, marked for cherry picking for 4.2 so it can be added to it for 4.2.1 if it doesn't meet the deadline

kitbdev commented:
Made an issue for the crash: https://github.com/godotengine/godot/issues/84125
If this isn't cherrypick-able I can create a smaller pr that is just the crash fix if needed.

YeldhamDev reviewed (approved):
Can't reproduce #66834 to confirm, but everything else seems alright.

YuriSizov commented:
I'm curious if @kitbdev could reproduce https://github.com/godotengine/godot/issues/66834, or if this is an assumed fix based on code changes.

YuriSizov commented on scene/gui/tab_container.cpp:
```diff
@@ -381,22 +381,29 @@ void TabContainer::move_tab_from_tab_container(TabContainer *p_from, int p_from_
 	Ref<Texture2D> tab_icon = p_from->get_tab_icon(p_from_index);
+	Ref<Texture2D> tab_button_icon = p_from->get_tab_button_icon(p_from_index);
 	bool tab_disabled = p_from->is_tab_disabled(p_from_index);
+	bool tab_hidden = p_from->is_tab_hidden(p_from_index);
 	Variant tab_metadata = p_from->get_tab_metadata(p_from_index);
+	int tab_icon_max_width = p_from->get_tab_bar()->get_tab_icon_max_width(p_from_index);
 
 	Control *moving_tabc = p_from->get_tab_control(p_from_index);
 	p_from->remove_child(moving_tabc);
 	add_child(moving_tabc, true);
 
-	set_tab_title(get_tab_count() - 1, tab_title);
-	set_tab_icon(get_tab_count() - 1, tab_icon);
-	set_tab_disabled(get_tab_count() - 1, tab_disabled);
-	set_tab_metadata(get_tab_count() - 1, tab_metadata);
-
 	if (p_to_index < 0 || p_to_index > get_tab_count() - 1) {
 		p_to_index = get_tab_count() - 1;
 	}
 	move_child(moving_tabc, get_tab_control(p_to_index)->get_index(false));
+
+	set_tab_title(p_to_index, tab_title);
+	set_tab_icon(p_to_index, tab_icon);
+	set_tab_button_icon(p_to_index, tab_button_icon);
+	set_tab_disabled(p_to_index, tab_disabled);
+	set_tab_hidden(p_to_index, tab_hidden);
+	set_tab_metadata(p_to_index, tab_metadata);
+	get_tab_bar()->set_tab_icon_max_width(p_to_index, tab_icon_max_width);
+
```
This isn't really maintainable. I think we should create a method in `TabBar` that returns a copy of the entire `Tab` struct and add API to `TabContainer` which can fetch it and apply it to one of its tabs. It doesn't need to be exposed to scripting, just public for internal use.

YuriSizov replied:
To be clear, this is not a requirement for this PR, but with the number of things we need to keep track of, it's a bigger concern now.

YuriSizov commented on editor/editor_node.cpp:
```diff
@@ -6593,6 +6513,10 @@ void EditorNode::_resource_loaded(Ref<Resource> p_resource, const String &p_path
 
 void EditorNode::_feature_profile_changed() {
 	Ref<EditorFeatureProfile> profile = feature_profile_manager->get_current_profile();
+	// Workaround: Close all floating docks to avoid crash.
```
Should be a FIXME comment.

YuriSizov commented on editor/editor_node.cpp:
```diff
@@ -5078,36 +5054,41 @@ void EditorNode::_update_dock_slots_visibility(bool p_keep_selected_tabs) {
 		for (int i = 0; i < vsplits.size(); i++) {
-			bool in_use = dock_slot[i * 2 + 0]->get_tab_count() || dock_slot[i * 2 + 1]->get_tab_count();
+			bool in_use = dock_slot[i * 2 + 0]->is_visible() || dock_slot[i * 2 + 1]->is_visible();
 			if (in_use) {
 				vsplits[i]->show();
 			} else {
 				vsplits[i]->hide();
 			}
```
This can be simplified with `set_visible`.

YuriSizov reviewed (approved):
Seems like a good refactoring. I'm not sure about the workaround for the crash. There are no other reports of it, it seems, and this is not a proper fix. So maybe it's not urgent to fix in 4.2 (so the entire PR can be pushed to 4.3, with maybe a cherry-pick later).

kitbdev commented:
I added some reproduction steps on https://github.com/godotengine/godot/issues/66834

I only found the crash since I was extracting all dock logic from `EditorNode` and found some issues and thought they should be applicable to 4.2, so I fixed them without the extraction.
I'm working on a different pr (open and close docks) that already has a proper fix for the crash.
So we can wait for that if we want to.

YuriSizov commented:
Since this PR is pretty much ready (just too late for 4.2), I think that we should merge it first, and then discuss your other PR built on top of this.

YuriSizov commented:
Cherry-picked for 4.2.2.

YuriSizov commented:
Cherry-picked for 4.1.4.
